_posts/2023-08-14-clear-specification-and-implementation.markdown

```diff
@@ -0,0 +1,610 @@
+show that multiple implementations, using different techniques, all
+fit the same notion of what it means to be a stack.
+
+To start with, let's specify what it means to be a stack in the
+following module.
+
+{% highlight dafny %}
+module StackSpecification {
+  // Most clients, including clients of implementations, get no details.
+  // This means that _all_ they know about stacks is what the two lemmas
+  // describe.
+  export
+    // The operations on a stack.
+    provides StackModel, PushModel, PopModel, EmptyModel, IsEmptyModel
+    // The properties of those operations.
+    provides PushPopVal, PushPopStack
+
+  // Implementation modules can access the definition of the model
+  // operations, to allow proofs of equivalence, by importing the
+  // `Private export set.
+  export Private
+    reveals StackModel, PushModel, PopModel, EmptyModel, IsEmptyModel
+    provides PushPopVal, PushPopStack
+
+  // We provide a concrete type to model a stack for simplicity of
+  // proof, since Dafny knows how to prove things about sequences. Note,
+  // however, that most clients don't get to see what this type is.
+  type StackModel<T> = seq<T>
+
+  // Check to see whether a stack is empty.
+  ghost function IsEmptyModel<T>(stk: StackModel<T>) : bool
+  {
+    |stk| == 0
+  }
+
+  // Construct an empty stack (and ensure that it's empty).
+  ghost function EmptyModel<T>() : StackModel<T>
+    ensures(IsEmptyModel(EmptyModel<T>()))
+  {
+    []
+  }
+
+  // Push an item onto the stack, ensuring that it's non-empty as a
+  // result.
+  ghost function PushModel<T>(stk: StackModel<T>, val : T) : StackModel<T>
+    ensures !IsEmptyModel(PushModel(stk, val))
+  {
+    [val] + stk
+  }
+
+  // Pop an item from a non-empty stack, yielding both an item and the
+  // rest of the stack.
+  ghost function PopModel<T>(stk: StackModel<T>) : (T, StackModel<T>)
+    requires !IsEmptyModel(stk)
+  {
+    (stk[0], stk[1..])
+  }
+
+  // The first core property of stacks, for use by clients. Pushing and
+  // then popping yields the pushed value.
+  lemma PushPopVal<T>(stk: StackModel<T>, val : T)
+    ensures PopModel(PushModel(stk, val)).0 == val {}
+
+  // The second core property of stacks, for use by clients. Pushing and
+  // then popping yields the original stack.
+  lemma PushPopStack<T>(stk: StackModel<T>, val : T)
+    ensures PopModel(PushModel(stk, val)).1 == stk {}
+}
+{% endhighlight %}
+
+Several things are important about this implementation. First, it
+provides a number of functions with bodies that describe one way of
+constructing a stack out of built-in primitives. Second, it hides
+the bodies of those functions from most clients, except
+implementation modules that need to prove equivalence with them.
+And, third, it describes the _relationship_ between those operations
+in a couple of lemmas.
+
+One further note: you'll see that some of the operations on a stack
+have pre- and post-conditions, specified on the functions themselves
```

Comment:
Maybe use the "intrinsic" vs "extrinsic" terminology? Maybe a good opportunity to say a few words about being mindful when choosing one vs the other?

Reply:
Yeah, I like the idea of using that terminology.